contrib/tools/meta.md/eqform.md

```diff
@@ -565,5 +579,146 @@ fmod EQFORM-SET is
+  eq normalize(M,DF \/ DG) = normalize(M,DF) \/ normalize(M,DG) .
+
+  eq normalize(M,T ?= T') = getTerm(metaNormalize(M,T)) ?= getTerm(metaNormalize(M,T')) .
+  eq normalize(M,T != T') = getTerm(metaNormalize(M,T)) != getTerm(metaNormalize(M,T')) .
+
+  op reduce : Module Form -> [Form] .
+  -----------------------------------
+  eq reduce(M,tt)       = tt .
+  eq reduce(M,ff)       = ff .
+  eq reduce(M,~ F)      = ~ reduce(M,F) .
+  eq reduce(M,CF /\ CG) = reduce(M,CF) /\ reduce(M,CG) .
+  eq reduce(M,DF \/ DG) = reduce(M,DF) \/ reduce(M,DG) .
+
+  eq reduce(M,T ?= T') = getTerm(metaReduce(M,T)) ?= getTerm(metaReduce(M,T')) .
+  eq reduce(M,T != T') = getTerm(metaReduce(M,T)) != getTerm(metaReduce(M,T')) .
+
+  op vars : Form -> QidSet .
+  --------------------------
+  eq vars(tt)       = none .
+  eq vars(ff)       = none .
+  eq vars(~ F)      = vars(F) .
+  eq vars(CF /\ CG) = vars(CF) ; vars(CG) .
+  eq vars(DF \/ DG) = vars(DF) ; vars(DG) .
+
+  eq vars(T ?= T') = vars(T) ; vars(T') .
+  eq vars(T != T') = vars(T) ; vars(T') .
+
+  --- INP: PosConj
+  --- PRE: PosConj has no ff literals
+  --- OUT: UnificationProblem
+  op toUnifProb : PosEqConj -> UnificationProblem .
+  ------------------------------------------------
+  eq toUnifProb((T ?= T') /\ PEC1) = T =? T' /\ toUnifProb(PEC1) .
+  eq toUnifProb(T ?= T')           = T =? T' .
+endfm
+```
+
+```maude
+fmod EQFORM-SET-OPERATIONS is
+  pr EQFORM-OPERATIONS .
+  pr EQFORM-SET .
+  pr EQFORM-CNF .
+  pr EQFORM-DNF .
+
+  var C : EqConj . var D : EqDisj . vars CF CG : NoTrueForm . vars DF DG : NoFalseForm .
+  var FS : FormSet . var F F' : Form . var UP : UnificationProblem .
+  var PEA : PosEqLit . var PELS : PosEqLitSet . var T T' : Term . var M : Module .
+
+  op wellFormed : Module FormSet -> Bool .
+  ----------------------------------------
+  eq wellFormed(M , F | F' | FS) = wellFormed(M,F) and-then wellFormed(M,F' | FS) .
+  eq wellFormed(M , mtFormSet)   = true .
+
+  op disj-join : FormSet -> [Form] .
+  op disj-join : FormSet -> [Form] .
+  ----------------------------------
+  eq disj-join(F | FS)    = F \/ disj-join(FS) .
+  eq disj-join(mtFormSet) = ff .
+
+  op conj-join : FormSet -> [Form] .
+  op conj-join : FormSet -> [Form] .
+  ----------------------------------
+  eq conj-join(F | FS)    = F /\ conj-join(FS) .
+  eq conj-join(mtFormSet) = tt .
+
+  op toDisjSet  : Form -> [EqDisjSet] .
+  op toDisjSet' : Form -> [EqDisjSet] .
+  -------------------------------------
+  eq toDisjSet (F)      = toDisjSet'(cnf(F)) .
+  eq toDisjSet'(tt)     = mtFormSet .
+  eq toDisjSet'(ff)     = mtFormSet .
+  eq toDisjSet'(D /\ F) = D | toDisjSet'(F) .
+
+  op toConjSet  : Form -> [EqConjSet] .
+  op toConjSet' : Form -> [EqConjSet] .
+  -------------------------------------
+  eq toConjSet (F)      = toConjSet'(dnf(F)) .
+  eq toConjSet'(tt)     = mtFormSet .
+  eq toConjSet'(ff)     = mtFormSet .
+  eq toConjSet'(C \/ F) = C | toConjSet'(F) .
```

Comment:
`toConjSet` and `toDisjSet` should probably not use top-sorted formula variables in recursion.

Reply:
I guess in this case, we're actually assuming that the input formula is a bunch of conjuncts. So perhaps we should leave the equation as-is, and instead change the typing to `EqConj -> [EqConjSet]`?

Reply:
As far as I know... the only difference between the top-sorted variables and the lowered ones is that the lowered variables omit `tt` or `ff` plus everything that is equivalent to `tt` or `ff` by identity, respectively. Thus, using the lowered sorts shouldn't change anything. Also, looking at this again, I realized we apply `dnf` first, which means that all formulas should be in normal equational form: either `tt` or `ff` or a bunch of equational atoms conjoined with `/\` and `\/`. So, we could even use some `Eq` sort probably.

Reply:
I misunderstood the change you were making earlier, but that's a good idea too. Other issues are just stylistic, so we can change it later if we want.